ctapipe/reco/hillas_intersection.py

```diff
@@ -103,7 +103,9 @@ def __call__(self, event):
         try:
             hillas_dict = self._create_hillas_dict(event)
         except (TooFewTelescopesException, InvalidWidthException):
-            return INVALID
+            event.dl2.stereo.geometry[self.__class__.__name__] = INVALID
```

Comment:
What I mean is that you could basically just return `None` and do nothing else and  `test_HillasReconstructor.py::test_invalid_events` still works.
So this test only checks if the code successfully runs into the exception and doesn't check if the code fills `INVALID` into the event structure.

Reply:
May be we can add something like `assert invalid_event.dl2.stereo.geometry` to the test to check if it was filled at all

Reply:
Yes, the test coverage here was clearly lacking, I will add a test for the invalid case

Reply:
this is again a case where the fact that Map is a defaultdict and not a dict basically made the test meaningless

Reply:
We should probably change that...

Reply:
What was the original reason for using a defaultdict instead of a dict?

Reply:
Convenience in creating the event data structure.
Instead of doing this:

```
event.dl1.tel[tel_id] = DL1CameraContainer()
event.dl1.tel[tel_id].image = ...
```

You could just do `event.dl1.tel[tel_id].image = ...`.

However, this has also worse performance, so we should just change it and add the explicit containers where needed.